cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/index.js

```diff
@@ -0,0 +1,323 @@
+            </div>
+        }
+        </div>
+      </div>
+    );
+  }
+
+  render() {
+    const actionLabel = T.translate('features.FastAction.setPreferencesLabel');
+    const headerTitle = `${actionLabel} for ${this.props.entity.id}`;
+    let tooltipID = `${this.props.entity.uniqueId}-setpreferences`;
+    return (
+      <span>
+        <FastActionButton
+          icon="fa fa-wrench"
+          action={this.toggleModal}
+          id={tooltipID}
+        />
+        <Tooltip
+          placement="top"
+          isOpen={this.state.tooltipOpen}
+          target={tooltipID}
+          toggle={this.toggleTooltip}
+          delay={0}
+        >
+          {T.translate('features.FastAction.setPreferencesLabel')}
+        </Tooltip>
+
+        {
+          this.state.modal ? (
+            <Modal
+              isOpen={this.state.modal}
+              toggle={this.toggleModal}
+              className="confirmation-modal"
+              size="lg"
+            >
+              <ModalHeader>
+                <div className="pull-left">
+                  <span
+                    className={"button-icon fa fa-wrench"}
+                  />
+                  &nbsp;{headerTitle}
+                </div>
+                <div className="pull-right">
+                  <div className="close-modal-btn"
+                    onClick={this.toggleModal.bind(this)}
+                  >
+                    <span
+                      className={"button-icon fa fa-times"}
+                    />
+                  </div>
+                </div>
+              </ModalHeader>
+              <ModalBody
+                className="modal-body"
+              >
+                <div className="preferences-container">
+                  <div className="specify-preferences-container">
+                    {this.renderSpecifyPreferences()}
+                    <div className="clearfix">
+                      {
+                        this.state.error ?
+                          <span className="pull-left text-danger"></span>
+                        :
+                          null
+                      }
+                      <button
+                        className="btn btn-primary pull-left"
+                        onClick={() => {this.toggleModal(); this.saveAppPreferences(); }}
+                      >
+                        <span>{T.translate('features.FastAction.setPreferencesButtonLabel')}</span>
+                      </button>
+                    </div>
+                  </div>
+                  <hr />
+                  <div className="inherited-preferences-container">
+                    {this.renderInheritedPreferences()}
+                  </div>
+                </div>
+              </ModalBody>
+            </Modal>
+          ) : null
+        }
+      </span>
+    );
+  }
+}
+
+SetPreferenceAction.propTypes = {
+  entity: PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    uniqueId: PropTypes.string,
```

Comment:
`id` here is actually the entity "name" (e.g. WordContent), while the `uniqueId` is randomly generated using shortId.generate(). I agree it's a bit confusing but all the other FastActions have both of them as props, so I didn't want to mess with them.